fix(serverless): adding ruby for containerized functions (#17410)
newrelic/docs-website: src/content/docs/serverless-function-monitoring/aws-lambda-monitoring/enable-lambda-monitoring/containerized-images.mdx

import serverlessContainerizedImagesWithLambda from 'images/serverless_diagram_containerized-images-with-lambda.webp'

If you're using a containerized image for a Lambda function and want to monitor your application, you'll need to add New Relic to your Dockerfile. For example, suppose you're a developer for a budgeting application and you're using a Lambda function to calculate account balances every time a customer clicks **See my balance**. You want to dive deep into any potential latencies. In that case, you'll need to add New Relic to your function, so every time a customer clicks **See my balance** your function runs, and New Relic does too.

> First, you have to add a reference to New Relic's prebuilt layer to your Dockerfile. From there, you'll need to create a new image. After you've created a new image with the reference to New Relic merged with your original Dockerfile, you'll push the image to your [AWS Elastic Container Registry](https://console.aws.amazon.com/ecr/home/). Now, you'll need to create a _new function_ from the image you pushed to AWS ECR.

Reviewer: @homelessbirds you might consider this for a subsequent revision, but consider summarizing these steps instead of spelling them all out. It's a bit of tough plowing to get through those in paragraph form. I'm not sure you want to make bullets since these are all listed below.  Even users with screen readers will see the steps below and sighted readers can scan the image to get the gist. You might say something like this: Here's a diagram showing the process of adding New Relic to the Dockerfile so you can monitor your function.

Author: This is good feedback. I've changed it.

    * [Java](https://gallery.ecr.aws/newrelic-lambda-layers-for-docker/newrelic-lambda-layers-java)
    * [Node.js](https://gallery.ecr.aws/newrelic-lambda-layers-for-docker/newrelic-lambda-layers-nodejs) 
    * [Python](https://gallery.ecr.aws/x6n7b2o2/newrelic-lambda-layers-python)
    * [Ruby](https://gallery.ecr.aws/x6n7b2o2/newrelic-lambda-layers-ruby)
    * [Extension-only](https://gallery.ecr.aws/newrelic-lambda-layers-for-docker/newrelic-lambda-layers-lambdaextension)
1. In your multi-stage Dockerfile, use New Relic's prebuilt image as an intermediate build stage. Then, copy New Relic's agent and extension from our prebuilt image into your final build stage.

> ### Example

Reviewer: @homelessbirds I didn't notice this before, but this Node.js example is different from what's in the tabs below. Why don't the examples in the tabs show the base image?

Author: @rhetoric101 the base image is dependent on what the user chooses. We can't pick it for them.

Author: Also, those are just examples-- they might be different for each person, but they just show what it could look like based on the most popular use case

        <TabsPageItem id="nodejs">

                ```dockerfile
                FROM public.ecr.aws/newrelic-lambda-layers-for-docker/newrelic-lambda-layers-nodejs:20-arm64 AS layer

                # Layer code
                COPY --from=layer /opt/ /opt/
                ```
        </TabsPageItem>

        <TabsPageItem id="Ruby">

                ```dockerfile

                FROM public.ecr.aws/newrelic-lambda-layers-for-docker/newrelic-lambda-layers-ruby:32-arm64 AS layer

                # Layer code
                COPY --from=layer /opt/ /opt/
                ```
        </TabsPageItem>

        <TabsPageItem id="extension">

                ```dockerfile
                FROM public.ecr.aws/newrelic-lambda-layers-for-docker/newrelic-lambda-layers-lambdaextension:x86 as layer

                # Layer code
                COPY --from=layer /opt/ /opt/
                ```
        </TabsPageItem>
    </TabsPages>
</Tabs>

</Step>
<Step>

### Create an image

>         1. After you add the reference to New Relic's pre-built image to your Dockerfile, you'll need to build a new image from the modified Dockerfile. This new image will have your original code _and_ New Relic's pre-built image. To do this, we recommend following the steps listed in the AWS guide, [Create a Lambda function using a container image](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html).

Reviewer: ```suggestion
        1. After you add the reference to New Relic's pre-built image to your Dockerfile, you'll need to build a new image from the modified Dockerfile. This new image will have your original code and New Relic's pre-built image. To do this, we recommend following the steps listed in the AWS guide, [Create a Lambda function using a container image](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html).
```

Reviewer: We try not to use italics.

                COPY --from=layer /opt/ /opt/
                ```
        </TabsPageItem>
    </TabsPages>
</Tabs>

</Step>
<Step>

### Create an image
        1. After you add the New Relic pre-built layer to your Dockerfile, you'll need to build a new image from the Dockerfile. This new image will have your original code _and_ New Relic's layer. To do this, we recommend following the steps listed in the AWS guide, [Pushing a Docker image](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html).
        2. Push the new image to your [AWS Elastic Container Registry](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html) repository for deployment.
</Step>

<Step>
### Configure your Lambda function

>         1. Create a new Lambda function using the image you pushed to AWS ECR. For more information on creating this function, we recommend reviewing the AWS guide: [Working with container images](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html).

Reviewer 2: ```suggestion
        1. Create a new Lambda function using the image you pushed to AWS ECR. For more information on creating this function, we recommend reviewing the AWS guide: [Pushing a Docker image to an Amazon ECR private repository](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html).
```

>         2. Configure the Lambda function with the same settings as your original function, but ensure you add any required environment variables for New Relic Lambda monitoring. See our [Environment variables](/docs/serverless-function-monitoring/aws-lambda-monitoring/enable-lambda-monitoring/containerized-images/#environment) for more information.

Reviewer 2: ```suggestion
        2. Configure the Lambda function with the same settings as your original function, but ensure you add any required environment variables for New Relic Lambda monitoring. See our [Environment variables](#environment) for more information.
```

        3. Update your function's handler to point to the newly attached layer in `cmd` override for your function. You can do that using the instructions in these two docs:
                * [Lambda console UI configuration](/docs/serverless-function-monitoring/aws-lambda-monitoring/enable-lambda-monitoring/instrument-your-own/#console)

>                 * [Layerless instrumentation for Lambda monitoring](docs/serverless-function-monitoring/aws-lambda-monitoring/enable-lambda-monitoring/enable-serverless-monitoring-aws-lambda-layerless)

Reviewer 2: ```suggestion
                * [Layerless instrumentation for Lambda monitoring](/docs/serverless-function-monitoring/aws-lambda-monitoring/enable-lambda-monitoring/enable-serverless-monitoring-aws-lambda-layerless)
```

        </tr>
    </tbody>
</table>

You can find more environment variables in our [Python configuration documentation](/docs/apm/agents/python-agent/configuration/python-agent-configuration).

</Collapser>

<Collapser
    id="java"
    title="Java OpenTracing"
>
<table>
    <thead>
        <tr>
            <th>Environment variable</th>
            <th>Default value</th>
            <th>Options</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`NEW_RELIC_PRIMARY_APPLICATION_ID`</td>
            <td></td>
            <td></td>
            <td>Your New Relic account ID</td>
        </tr>
        <tr>
            <td>`NEW_RELIC_DEBUG`</td>
            <td>`false`</td>
            <td>`true`, `false`</td>
            <td>Agent log level</td>
        </tr>
    </tbody>
</table>

</Collapser>

<Collapser
    id="extension"
    title="Extension"
>

<table>
    <thead>
        <tr>
            <th>Environment variable</th>
            <th>Default value</th>
            <th>Options</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>

> <td>`NEW_RELIC_LAMBDA_EXTENSION_ENABLED`</td>

Author: it looks correct on my end?
